core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -154,6 +154,10 @@ int HIPInternal::verify_is_initialized(const char *const label) const {
   return 0 <= m_hipDev;
 }
 
+uint32_t HIPInternal::impl_get_instance_id() const noexcept {
+  return Kokkos::Tools::Experimental::Impl::idForInstance<
+      Kokkos::Experimental::HIP>(reinterpret_cast<uintptr_t>(this));
```

Comment:
Could you please elaborate on why you need all this magic here? In which cases do you want the same id to be returned? Does this work the same way for all backends?

Reply:
`Does this work the same way for all backends?` For all backends which have distinct instances, it works this way. Serial doesn't work this way, because it only has the one instance.

I need all of this magic because for distinguishable instances, I need to map those instances to an ID. I want the same ID to be returned if the same instance is used. So if somebody creates a CUDA/HIP instance from different streams, we would want to see different ID's here.

Reply:
I was slightly confused that using the address to create a hash is what we want. Of course, this works because these classes are always stored as pointers in the execution space classes.